lib/runner.js

```diff
@@ -529,8 +529,12 @@ Runner.prototype.runTests = function(suite, fn) {
         test = self.test;
 
         if (err) {
+          var retry = test.currentRetry();
           if (err instanceof Pending) {
             self.emit('pending', test);
+          } else if (retry < test.retries()) {
+            test.currentRetry(retry + 1);
+            tests.unshift(test);
```

Comment:
Given `var tests = suite.tests.slice();`, adding it to `tests` is a fine approach. As long as `suite.tests` isn't being modified :)